### PR TITLE
Multistage Docker Builds

### LINKED ISSUE #129: OSIE image should use multi stage builds
We build a few packages from source, some of them may go away with #128 but some are likely to still be needed. Using multi stage builds would allow making better use layer caching and likely help with the final image size (#2).

### PR DESCRIPTION
**Pull request by mmlb**
## Description

Changes the osie container to use multistage builds for every binary built from source.

## Why is this needed

Make builds faster. Fixes #129

## How Has This Been Tested?

Builds run, all the tools are there and can be invoked.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,34 @@
-FROM ubuntu:xenial
+FROM ubuntu:xenial as builder
```
Comment (markjacksonfishing) [inline docker/Dockerfile:1]: Should we be pinning a version here?

File: docker/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,34 @@
...
+RUN apt-get update -y && apt-get install -y build-essential curl
```
Comment (markjacksonfishing) [inline docker/Dockerfile:2]: `--no-install-recommends` should also be in here

File: docker/Dockerfile [hunk 3]
```diff
@@ -1,4 +1,34 @@
...
+
+FROM builder as racadm
+RUN apt-get update && apt-get install -y alien wget
+RUN rpm --verbose --import https://linux.dell.com/repo/pgp_pubkeys/0x1285491434D8786F.asc
+WORKDIR /tmp/osie
+RUN wget \
+        https://dl.dell.com/FOLDER05920767M/1/DellEMC-iDRACTools-Web-LX-9.4.0-3732_A00.tar.gz \
+        http://linux.dell.com/repo/community/openmanage/940/bionic/pool/main/s/srvadmin-omilcore/srvadmin-omilcore_9.4.0_amd64.deb
+RUN tar -xvf DellEMC-iDRACTools-Web-LX-9.4.0-3732_A00.tar.gz
+RUN alien iDRACTools/racadm/RHEL8/x86_64/*.rpm
```
Comment (markjacksonfishing) [inline docker/Dockerfile]: Use ./*glob* or -- *glob* so names with dashes won't become options.

**Review by markjacksonfishing — CHANGES_REQUESTED** [review]
I know this is still in draft but some initial comments

File: docker/Dockerfile (hunk 3 above)
Comment (mmlb) [reply docker/Dockerfile]: that doesn't apply hear since the `iDrac...` does the same effect as `./`

File: docker/Dockerfile (hunk 2 above)
Comment (mmlb) [reply docker/Dockerfile:2]: I'll give this a try.

File: docker/Dockerfile (hunk 1 above)
Comment (mmlb) [reply docker/Dockerfile:1]: I don't think its necessary. Doesn't buy us much. The apt caches are still blown away so we'd still need to do an `apt-get update` for most of the things we want. I've got a jira ticket to possibly explore pinning the version to a date and what not but that'll be left for another day.

File: docker/Dockerfile (hunk 2 above)
Comment (mmlb) [reply docker/Dockerfile:2]: Nope that doesn't work, and seeing as this is just the build image I'm ok with having some extra junk in there.

**Review by markjacksonfishing — DISMISSED** [review]

**Comment by mmlb** [comment]
This is in the back burner for the moment, need to figure out how to squash the `buildx` built images.

**Comment by nshalman** [comment]
@mmlb should we keep this open?

**Comment by jacobweinstock** [comment]
Discussed this with @mmlb in the Triage party and he will re-open this if/when desired/needed.

**Pull request closed without merging**